### PR TITLE
hyphen-case config key support

### PR DESCRIPTION
**Pull request by shanielh**
Just a rebase of https://github.com/ceedubs/ficus/pull/12 to iheartradio/ficus

### CODE REVIEW

**Comment by kailuowang** [comment]
@shanielh thanks for the rebase. I had the same thoughts as @ceedubs (the guava dependency). Also we have the plan to rewrite the automatic case class readers based on shapeless. I am working on some [shapeless+cat facilities](https://github.com/milessabin/kittens) that will make even more easier. Hopefully we'll see that soon. So for now I am going to leave this one on hold for a bit. Please check back later.

File: build.sbt [hunk 1]
```diff
@@ -5,6 +5,9 @@ description := "A Scala-friendly wrapper companion for Typesafe config"
 
 startYear := Some(2013)
 
+licenses := Seq(
+  "MIT License" -> url("http://www.opensource.org/licenses/mit-license.html")
+)
```
Comment (kailuowang) [inline build.sbt]: Not needed, here, license is moved to project/Publish.scala

**Comment by ceedubs** [comment]
Travis Brown has done some [recent work](https://meta.plasm.us/posts/2016/01/14/configuring-generic-derivation/) support different cases in Circe. Since it uses Shapeless for derivation, it might be a good reference.

**Comment by kailuowang** [comment]
@ceedubs, that looks quite similar, thanks for the hint.

File: src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala [hunk 2]
```diff
@@ -11,22 +12,48 @@ trait ArbitraryTypeReader {
...
 object ArbitraryTypeReaderMacros {
   import scala.reflect.macros.blackbox.Context
 
-  def arbitraryTypeValueReader[T : c.WeakTypeTag](c: Context): c.Expr[ValueReader[T]] = {
+  trait NameMapper {
+    def map(name: String): String
+  }
+
+  object defaultMapper extends NameMapper {
+    def map(name: String) = name
```
Comment (kailuowang) [inline src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala]: Codacy Issue found: [Public method must have explicit type](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786830/issues/source?bid=3018830&fileBranchId=3073710#l29)

File: src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala [hunk 3]
```diff
@@ -11,22 +12,48 @@ trait ArbitraryTypeReader {
...
+  }
+
+  object hyphenCaseMapper extends NameMapper {
```
Comment (kailuowang) [inline src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala]: Codacy Issue found: [Object name does not match the regular expression '^[A-Z][A-Za-z0-9]*$'](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786830/issues/source?bid=3018830&fileBranchId=3073710#l32)

File: src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala [hunk 4]
```diff
@@ -49,7 +76,7 @@ object ArbitraryTypeReaderMacros {
     c.Expr[T](Apply(instantiationCall, instantiationArgs))
   }
 
-  def extractMethodArgsFromConfig[T : c.WeakTypeTag](c: Context)(method: c.universe.MethodSymbol, companionObjectMaybe: Option[c.Symbol],
+  def extractMethodArgsFromConfig[T : c.WeakTypeTag](c: Context, paramMapper: NameMapper)(method: c.universe.MethodSymbol, companionObjectMaybe: Option[c.Symbol],
```
Comment (kailuowang) [inline src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala]: Codacy Issue found: [File line length exceeds 160 characters](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786830/issues/source?bid=3018830&fileBranchId=3073710#l79)

File: src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala [hunk 5]
```diff
@@ -11,22 +12,48 @@ trait ArbitraryTypeReader {
...
+
 object ArbitraryTypeReaderMacros {
   import scala.reflect.macros.blackbox.Context
 
-  def arbitraryTypeValueReader[T : c.WeakTypeTag](c: Context): c.Expr[ValueReader[T]] = {
+  trait NameMapper {
+    def map(name: String): String
+  }
+
+  object defaultMapper extends NameMapper {
```
Comment (kailuowang) [inline src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala]: Codacy Issue found: [Object name does not match the regular expression '^[A-Z][A-Za-z0-9]*$'](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786830/issues/source?bid=3018830&fileBranchId=3073710#l28)

File: src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala [hunk 6]
```diff
@@ -11,22 +12,48 @@ trait ArbitraryTypeReader {
...
+    def map(name: String) = name
+  }
+
+  object hyphenCaseMapper extends NameMapper {
+    def map(name: String) = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_HYPHEN, name)
```
Comment (kailuowang) [inline src/main/scala/net/ceedubs/ficus/readers/ArbitraryTypeReader.scala]: Codacy Issue found: [Public method must have explicit type](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786830/issues/source?bid=3018830&fileBranchId=3073710#l33)

File: src/test/scala/net/ceedubs/ficus/readers/ArbitraryTypeReaderSpec.scala [hunk 7]
```diff
@@ -85,6 +86,14 @@ class ArbitraryTypeReaderSpec extends Spec { def is = s2"""
     instance.foo must_== foo
   }
 
+  def withCamelCaseFields = {
```
Comment (kailuowang) [inline src/test/scala/net/ceedubs/ficus/readers/ArbitraryTypeReaderSpec.scala]: Codacy Issue found: [Public method must have explicit type](https://www.codacy.com/app/kailuo-wang/ficus/file/1805786847/issues/source?bid=3018830&fileBranchId=3073710#l89)

**Comment by larroy** [comment]
Please don't add huge dependencies such as Guava.

**Comment by colindean** [comment]
Hyphen case would be super useful for my team. What work needs to be done to get this into Ficus?

**Comment by kailuowang** [comment]
@colindean the plan was to take advantage of [henkan](https://github.com/kailuowang/henkan) to revamp the currently macro based auto case class mapping. However henkan is having a weird bug https://github.com/kailuowang/henkan/issues/15. I would like to take advantage of @milessabin's [SI2712 fix](https://github.com/scala/scala/pull/5102) to simplify henkan before jumping into that bug fix though...
so it might take a while.

**Comment by colindean** [comment]
Ah. I guess we'll stick to snake case for now!

**Comment by kailuowang** [comment]
Okay. I gave up on using henkans here. So for this PR we just need to replace `CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_HYPHEN, name` 
with a hand-roll implementation and get rid of the guava dependency, then we can merge.

**Comment by shanielh** [comment]
@kailuowang Great! I can do the changes, But before I do them, Maybe I should expose the `NameMapper` trait and get it in the arbitrary type reader as an implicit parameter? That way everyone can define their own `NameMapper`. In that case we can get rid of the guava dependency and even not implement an `HypenCaseNameMapper`.

**Comment by joprice** [comment]
@shanielh that's the exact conversation that Kai and I just had about this design. I think that's cleaner and more extensible.

**Comment by shanielh** [comment]
@joprice Hah, Ok, I believe I'll fix the PR tomorrow 👍

**Comment by kailuowang** [comment]
@shanielh thanks very much!

**Comment by shanielh** [comment]
Continued in #22 👍

**Pull request closed without merging**